### PR TITLE
perf(sql): optimize like/ilike/regexp functions on symbol column

### LINKED ISSUE #4825: Symbol like/ilike/regexp filters may be slow when the number of distinct values is high
### To reproduce

In v8.1 we introduced optimized functions for `like`/`ilike`/`~` operators on symbol columns. Those operators first filter the symbol table, store a list of `int` codes, and then use it to filter the rows. It turns out that this may be slower than the old string function in case of a large number of distinct symbols.

We should make this optimization optional, depending on the known size of symbol table.

### QuestDB version:

8.1

### OS, in case of Docker specify Docker and the Host OS:

Linux

### File System, in case of Docker specify Host File System:

ext4

### Full Name:

Andrei Pechkurov

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Fixes #4825

The bottleneck was in the expensive `StatefulAtom#init()` calls for the `AbstractLikeSymbolFunctionFactory` and `MatchSymbolFunctionFactory` functions. These calls were made for all per-worker filters on a single thread. See `AsyncFilterAtom` as an example: https://github.com/questdb/questdb/blob/fba356e67ae8b3ce77bfff2b68a1a83c86da63ad/core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java#L99

So, the more worker threads the server had and the more symbols were in the symbol table, the more time was required for the serial part of the query execution, i.e. the `init()` calls.

Also, introduces additional functions for simple like/ilike patterns to `AbstractLikeSymbolFunctionFactory` - that's to speed up symbol table iteration.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 159 / 194 (81.96%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/AbstractLikeSymbolFunctionFactory.java|150|185|81.08%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchSymbolFunctionFactory.java|9|9|100.00%|

**Comment by Abhiram98** [comment]
Hi! I left a few comments about making some cosmetic improvements to the code, mainly to the "newInstance" function, as it was a bit hard to follow. 
I am happy to send a patch with some/all of the suggested changes. Let me know what you think!

**Review by ideoma — COMMENTED** [review]
Can you please make the coverage 100%? I suspect one of the optimisation branches is not covered.
Also, consider this pattern

select * from x where sym like '%\%'

It will go into "contains" branch but it should be the "starts with" branch.

**Comment by puzpuzpuz** [comment]
> Can you please make the coverage 100%? I suspect one of the optimisation branches is not covered.
> Also, consider this pattern

Fixed the pattern handling and added missing tests in 53caed0ae2e91665b24e2dae538f5f663c77f924. PTAL

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 163 / 195 (83.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/AbstractLikeSymbolFunctionFactory.java|150|182|82.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchSymbolFunctionFactory.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/AbstractLikeVarcharFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/AbstractLikeStrFunctionFactory.java|2|2|100.00%|

**Review by ideoma — APPROVED** [review]
